### ISSUE TITLE
Determine command pattern

### ISSUE THREAD

**Issue by CS-5** [Team4050/2023-charged-up#8]
Need to figure out what makes sense for a general pattern to follow when creating commands:
- What justifies the use of a standalone command (as opposed to inline commands in RobotContainer)?
- How should commands be separated? 
  - e.g.: "Scoring" command (for arm and claw) vs distinct "Claw" and "Arm" commands
- Should controller input be handled inside or outside of commands
  - DanceCommand is an example of input being outside of the command
  - ClawToggleCmd is an example of input being inside the command
- Other ideas?

**Comment by Kieth-the-25th** [comment]
Another thing which relates to this is abstracting the controls a bit so we could change them easily, and figuring out when to use the Trigger class or have conditional stuff.

**Comment by CS-5** [comment]
Agreed, great point! I would lean towards having all controller button/axis mappings in RobotContainer, but there are other ways to accomplish the same thing too.

**Comment by Kieth-the-25th** [comment]
@CS-5 So I've spent some time thinking, and I think this is reasonable:

- Commands with no methods other than execute() and initialize() and no stored value types should be inline (because Runnables physically can't store value types.)
- Each command should access only one subsystem and commands shouldn't need to communicate with each other, so we can have all our control structure in one place.
- Personally, I think it makes more sense for button-mapped commands to handle input outside of the command (like ClawToggleCmd being bound to a Trigger object), and for commands needing joystick input to handle that inside the command (like our drivetrain inline command), since button mappings will probably change more than joystick mappings.
- Additionally, commands should have different "modes" rather than a separate command for each "mode" (e.g. a separate command for each arm height.) Personally, I think it's just easier to write it that way.

**Comment by CS-5** [comment]
Great ideas, I agree on all points :+1:

On your last point, I think there may be cases in very complex robots where splitting code into multiple separate commands might make sense, but I don't believe we're there yet.

---

We will want to see how autonomous changes things (if at all), but I say lets move forward with the above. Nice work!

**Comment by Kieth-the-25th** [comment]
Autonomous will definitely change things, since in autonomous we'll likely have to use multiple subsystems in sync, and a single command is the easiest way of doing that.
